Sources/GRPCHTTP2Core/Client/Connection/Connection.swift

```diff
@@ -0,0 +1,434 @@
+        // Fire the error inbound; this fails the inbound writer.
+        self.http2Stream.channel.pipeline.fireErrorCaught(error)
+      }
+    }
+
+    let descriptor: MethodDescriptor
+
+    private let http2Stream: NIOAsyncChannel<RPCResponsePart, RPCRequestPart>
+
+    init(
+      wrapping stream: NIOAsyncChannel<RPCResponsePart, RPCRequestPart>,
+      descriptor: MethodDescriptor
+    ) {
+      self.http2Stream = stream
+      self.descriptor = descriptor
+    }
+
+    func execute<T>(
+      _ closure: (_ inbound: Inbound, _ outbound: Outbound) async throws -> T
+    ) async throws -> T where T: Sendable {
+      try await self.http2Stream.executeThenClose { inbound, outbound in
+        return try await closure(
+          inbound,
+          Outbound(requestWriter: outbound, http2Stream: self.http2Stream)
+        )
+      }
+    }
+  }
+}
+
+@available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+extension Connection {
+  private enum State {
+    /// The connection is idle or connecting.
+    case notConnected
+    /// A connection has been established with the remote peer.
+    case connected(Connected)
+    /// The connection has started to close. This may be initiated locally or by the remote.
+    case closing
+    /// The connection has closed. This is a terminal state.
+    case closed
+
+    struct Connected {
+      /// The connection channel.
+      var channel: NIOAsyncChannel<ClientConnectionEvent, Void>
+      /// Multiplexer for creating HTTP/2 streams.
+      var multiplexer: NIOHTTP2Handler.AsyncStreamMultiplexer<Void>
+      /// Whether the connection is plaintext, `false` implies TLS is being used.
+      var scheme: Scheme
+
+      init(_ connection: HTTP2Connection) {
+        self.channel = connection.channel
+        self.multiplexer = connection.multiplexer
+        self.scheme = connection.isPlaintext ? .http : .https
+      }
+    }
+
+    mutating func connected(_ channel: HTTP2Connection) {
+      switch self {
+      case .notConnected:
+        self = .connected(State.Connected(channel))
+      case .connected, .closing, .closed:
+        fatalError("Invalid state: 'run()' must only be called once")
+      }
+    }
+
+    mutating func beginClosing() -> NIOAsyncChannel<ClientConnectionEvent, Void>? {
+      switch self {
+      case .notConnected:
+        fatalError("Invalid state: 'run()' must be called first")
+      case .connected(let connected):
+        self = .closing
+        return connected.channel
+      case .closing, .closed:
+        return nil
+      }
+    }
+
+    mutating func closing() {
+      switch self {
+      case .notConnected:
+        // Not reachable: happens as a result of a connection event, that can only happen if
+        // the connection has started (i.e. must be in the 'connected' state or later).
+        fatalError("Invalid state")
+      case .connected:
+        self = .closing
+      case .closing, .closed:
+        ()
+      }
+    }
```

Comment:
I think we do, they're semantically different. `beginClosing` is something we ask the connection to do, `closing` is in response to the connection closing itself (idle timeout, keaplive failed etc.). 

I think the semantic difference is important here. It would also mean in places where `closing()` is called we'd have to handle the async channel being returned which should never happen.